packages/excalidraw/components/App.tsx

```diff
@@ -7993,12 +7993,35 @@ class App extends React.Component<AppProps, AppState> {
 
         if (newElement.type === "freedraw") {
           const points = newElement.points;
-          const dx = pointerCoords.x - newElement.x;
-          const dy = pointerCoords.y - newElement.y;
+          let dx = pointerCoords.x - newElement.x;
+          let dy = pointerCoords.y - newElement.y;
+
+          if(event.shiftKey) {
+
+            const angle = Math.atan2(dy, dx);
+            const distance = Math.sqrt(dx * dx + dy * dy);
+
+            const snappingAngles = [
+              0, Math.PI / 6, Math.PI / 4, Math.PI / 3, Math.PI / 2,
+              (2 * Math.PI) / 3, (3 * Math.PI) / 4, (5 * Math.PI) / 6,
+              Math.PI, (7 * Math.PI) / 6, (5 * Math.PI) / 4, (4 * Math.PI) / 3,
+              (3 * Math.PI) / 2, (5 * Math.PI) / 3, (7 * Math.PI) / 4, (11 * Math.PI) / 6
+            ];
+            
+            const closestAngle = snappingAngles.reduce((prev, curr) =>
+              Math.abs(curr - angle) < Math.abs(prev - angle) ? curr : prev
+            );
```

Comment:
This should be a binary search instead of an exhaustive linear search, there is no such function yet, but it could go into `packages/utils/binarysearch.ts`.